tests/std/tests/P2321R2_views_zip/env.lst

```diff
@@ -0,0 +1,4 @@
+# Copyright (c) Microsoft Corporation.
+# SPDX-License-Identifier: Apache-2.0 WITH LLVM-exception
+
+RUNALL_INCLUDE ..\strict_concepts_latest_matrix.lst

```

Comment:
Hmm... I was using `strict_concepts_latest_matrix.lst` instead of `concepts_latest_matrix.lst` because using the latter was causing the test suite to fail to compile. Is there anything particularly wrong with this?

Reply:
I've done some analysis, and I'm inclined to believe that something is not quite right here. Notice that `zip_view::_Iterator` has an overload for `iter_move` and `iter_swap`. Every other view type in `<ranges>` which overloads these functions for use in ADL also uses `strict_concepts_latest_matrix.lst`, rather than `concepts_latest_matrix.lst` (or, analogously, `strict_concepts_20_matrix.lst`).

The interesting thing is that my test isn't the only one to fail to compile if one changes `strict_concepts_latest_matrix.lst` to `concepts_latest_matrix.lst`. Take, for instance, the test for `P2441R2_views_join_with`. Note that `join_with_view::_Iterator` has its own overloads for `iter_move` and `iter_swap`. When using `concepts_latest_matrix.lst` instead of `strict_concepts_latest_matrix.lst`, this test fails to compile with the same error as mine when `/permissive` is used instead of `/permissive-`: The compiler throws an error for an ambiguous symbol lookup on `iter_move`. (This particular test doesn't verify `iter_swap`, although `join_with_view::_Iterator` does define an overload for it.)

With all of this said, I am left with a few questions. Is it really wrong to be using `strict_concepts_latest_matrix.lst` instead of `concepts_latest_matrix.lst`? If so, then why should the other tests be exempt from this? Shouldn't they also be using `concepts_latest_matrix.lst`? For now, I am going to keep using `strict_concepts_latest_matrix.lst`, but I feel that this issue warrants further discussion.

Reply:
We've been cracking down recently on tests using `strict_meow.lst`s that pass just fine with the normal `meow.lst`. I suspect Nicole's (should I call you "Nicole" or "striga" or "@striga-nil-ms" or "ubsan" in GH PRs?) intent here was to verify that you had tried the non-strict matrix and seen a failure.

> I've done some analysis, and I'm inclined to believe that something is not quite right here. Notice that `zip_view::_Iterator` has an overload for `iter_move` and `iter_swap`. Every other view type in `<ranges>` which overloads these functions for use in ADL also uses `strict_concepts_latest_matrix.lst`, rather than `concepts_latest_matrix.lst` (or, analogously, `strict_concepts_20_matrix.lst`).

Winner! This is the primary problem Ranges has with MSVC permissive mode. MSVC's permissive mode name lookup falls flat on its face when we have friend functions ("hidden" friends aren't hidden in permissive mode!) appear in a namespace when an object with the same name appears in the inline namespace set. Consequently all iterator types that customize `iter_move` and/or `iter_swap` whose nearest enclosing namespace is `std::ranges` need the strict matrix.

Reply:
Thanks for the explanation! Sadly, that's quite the unfortunate situation to be in. Strictly speaking, we couldn't just *not* make `std::ranges::iter_move` and `std::ranges::iter_swap` customization point objects, because that's required by the standard. By making them CPOs, however, we can't reliably define similarly named functions in the same namespace when `/permissive` is being used.

Well, now I know why `/permissive-` is set by default when using `/std:c++20` or later (or at least, one of the reasons why).

Reply:
@CaseyCarter Nicole or strega, ubsan is an old nick I used to use but no longer! 😝

Reply:
> Sadly, that's quite the unfortunate situation to be in.

I'm perfectly happy telling folks they need to "delegacy" their codebase to fully enjoy the benefits of modern C++.